### PR TITLE
Correctness

### PR DESCRIPTION
**Pull request by friendlyanon**
Since this thing sees some use in the wild, I decided to fix some correctness issues.

I wrote the rationale in the commit messages where applicable, but ask away if you have any questions.

### CODE REVIEW

File: CMakeLists.txt
Comment (TheLartians) [inline CMakeLists.txt:74]: Interesting, is it documented behaviour that setting `CMAKE_SIZEOF_VOID_P` to an empty string disables the arch check? Would be great if there was some documentation to this behaviour so we can assume that it won't break between CMake versions.

If I understand correctly, the reason we do this is that we currently do not set a minimum CMake version. Do you think it would be better to define the minimum version to `3.14` and pass the `PROJECT_ARCH_INDEPENDENT` to `write_basic_package_version_file` instead?

**Review by TheLartians — COMMENTED** [review]
Hey, thanks for the PR @friendlyanon! I agree that it does look more correct this way. One thing that I don't quite get is why changing `CMAKE_SIZEOF_VOID_P` would be equivalent to setting `ARCH_INDEPENDENT`. Is this documented somewhere?

File: CMakeLists.txt
Comment (friendlyanon) [reply CMakeLists.txt:74]: You can see the code [here](https://github.com/Kitware/CMake/blame/master/Modules/BasicConfigVersion-SameMajorVersion.cmake.in#L57) for this. It hasn't changed in over a decade and `ARCH_INDEPENDENT` basically just does the same thing.

With CMake 3.14, this "hack" can be dropped and the namelink can be created with `NAMELINK_COMPONENT` instead, so that would simplify at least 2 things.

Comment (TheLartians) [reply CMakeLists.txt:74]: From the code I agree that it does look like the "intended undocumented" way of doing this.

I do think it should be ok to target CMake 3.14 as a minimum though, as even `ubuntu-latest` on GitHub CI seems to support it, and most users of this project should be able to use a fairly modern toolchain as well. If you agree, I'd prefer to use the clean version of setting a minimum and using the `PROJECT_ARCH_INDEPENDENT` feature as documented.

**Review by TheLartians — APPROVED** [review]
Looks great, thanks for the changes!

**Comment by friendlyanon** [comment]
Somehow 2 components names were left as is amidst all these rebases to fix formatting. I swear I can git really good.

File: CMakeLists.txt
Comment (friendlyanon) [inline CMakeLists.txt:69]: Actually, since this is [advertised](https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html#command:write_basic_package_version_file) as a feature mainly for header only libraries, it would be safe to detect the target type and default to using this.

Comment (TheLartians) [reply CMakeLists.txt:76]: Yeah the only situation where I can imagine that this goes wrong is if a library ships completely different headers for difference architectures. But afaik that would be exceptionally rare.

**Comment by TheLartians** [comment]
@friendlyanon is the PR ready for final review / merge? 
Also it's kinda hard to review incremental changes if you force-push, as the history might not be conserved.

File: CMakeLists.txt
Comment (friendlyanon) [reply CMakeLists.txt:76]: Yes, I think the general use-case is not that, so providing a sane default helps here.

**Comment by friendlyanon** [comment]
Yes, I don't think I have anything more to add.

> it's kinda hard to review incremental changes if you force-push, as the history might not be conserved

Hm, can you elaborate? git preserves the topological order of the commits when things get rebased. rebasing also helps with keeping the commit history clean.

**Review by TheLartians — APPROVED** [review]

**Comment by TheLartians** [comment]
> Hm, can you elaborate? git preserves the topological order of the commits when things get rebased. rebasing also helps with keeping the commit history clean.

For reviewing on GitHub I can't immediately see which commits are new / modified after a force push. Usually I'd just expect to view the new commits since the last time I viewed the PR, so there is less code to go through.

Regarding the history, the single commits and detailed messages are definitely great and very clean, but imo could also be detailed as an initial comment in the PR, as I tend to review the changes in their entirety anyways. Note that maintainers can  squash PRs into a single commit before merging from the GitHub UI, so that the history in the `main` branch can remain clean either way.

**Comment by TheLartians** [comment]
The changes have been released in `v1.6.0` 🎉

**Pull request merged**
